### PR TITLE
Avoid copies in `CountWildcardRule` via TreeNode API

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Part of https://github.com/apache/arrow-datafusion/issues/9637

## Rationale for this change

Let's make planning faster by not copying so much in the optimizer using the new TreeNode API.

As an added bonus, now that the traversal is encoded in `TreeNode` doing this replacement also requires much less code
## What changes are included in this PR?

1. See PR subject

## Are these changes tested?
Covered by existing CI

## Are there any user-facing changes?
No (maybe tiny bit faster planning)

### CODE REVIEW

File: datafusion/optimizer/src/analyzer/count_wildcard_rule.rs [hunk 1]
```diff
@@ -47,181 +38,43 @@ impl CountWildcardRule {
...
-                        },
-                        negated,
-                    })
-                }),
-            _ => Transformed::no(old_expr),
-        })
-    }
+fn analyze_internal(plan: LogicalPlan) -> Result<Transformed<LogicalPlan>> {
+    let name_preserver = NamePreserver::new(&plan);
+    plan.map_expressions(|expr| {
```
Comment (alamb) [inline datafusion/optimizer/src/analyzer/count_wildcard_rule.rs:55]: Not only does this logic avoid the clone (`old_expr.clone()`) I also think it is easier to understand what is happening

File: datafusion/optimizer/src/analyzer/count_wildcard_rule.rs [hunk 2]
```diff
@@ -47,181 +38,43 @@ impl CountWildcardRule {
 
 impl AnalyzerRule for CountWildcardRule {
     fn analyze(&self, plan: LogicalPlan, _: &ConfigOptions) -> Result<LogicalPlan> {
-        plan.transform_down(&analyze_internal).data()
+        plan.transform_down_with_subqueries(&analyze_internal)
```
Comment (alamb) [inline datafusion/optimizer/src/analyzer/count_wildcard_rule.rs:41]: By using the wonderful `transform_down_with_subqueries` from @peter-toth this rule can avoid having to recuse into subqueries directly by itself

File: datafusion/optimizer/src/analyzer/count_wildcard_rule.rs [hunk 3]
```diff
@@ -47,181 +38,43 @@ impl CountWildcardRule {
...
 }
 
-fn analyze_internal(plan: LogicalPlan) -> Result<Transformed<LogicalPlan>> {
-    let mut rewriter = CountWildcardRewriter {};
-    match plan {
-        LogicalPlan::Window(window) => {
-            let window_expr = window
-                .window_expr
-                .iter()
-                .map(|expr| rewrite_preserving_name(expr.clone(), &mut rewriter))
```
Comment (alamb) [inline datafusion/optimizer/src/analyzer/count_wildcard_rule.rs:65]: removed this clone (and the ones below)

File: datafusion/optimizer/src/analyzer/count_wildcard_rule.rs [hunk 4]
```diff
@@ -47,181 +38,43 @@ impl CountWildcardRule {
...
+    matches!(expr, Expr::Wildcard { qualifier: None })
 }
-
-struct CountWildcardRewriter {}
-
-impl TreeNodeRewriter for CountWildcardRewriter {
-    type Node = Expr;
-
-    fn f_up(&mut self, old_expr: Expr) -> Result<Transformed<Expr>> {
-        Ok(match old_expr.clone() {
```
Comment (alamb) [inline datafusion/optimizer/src/analyzer/count_wildcard_rule.rs:125]: here is another clone that is avoided

File: datafusion/optimizer/src/analyzer/function_rewrite.rs [hunk 5]
```diff
@@ -64,7 +64,7 @@ impl ApplyFunctionRewrites {
             let original_name = name_preserver.save(&expr)?;
 
             // recursively transform the expression, applying the rewrites at each step
-            let result = expr.transform_up(&|expr| {
```
Comment (alamb) [inline datafusion/optimizer/src/analyzer/function_rewrite.rs:67]: drive by cleanup based on @crepererum 's comment on https://github.com/apache/arrow-datafusion/pull/10038#discussion_r1560648005

File: datafusion/optimizer/src/analyzer/count_wildcard_rule.rs [hunk 6]
```diff
@@ -47,181 +38,43 @@ impl CountWildcardRule {
...
-        })
-    }
+fn analyze_internal(plan: LogicalPlan) -> Result<Transformed<LogicalPlan>> {
+    let name_preserver = NamePreserver::new(&plan);
+    plan.map_expressions(|expr| {
+        let original_name = name_preserver.save(&expr)?;
+        let transformed_expr = expr.transform_up(&|expr| match expr {
+            Expr::WindowFunction(mut window_function)
+                if window_function.args.len() == 1
+                    && is_wildcard(&window_function.args[0]) =>
```
Comment (jayzhan211) [inline datafusion/optimizer/src/analyzer/count_wildcard_rule.rs:60]: do we need to match function Count as well?

Comment (alamb) [reply datafusion/optimizer/src/analyzer/count_wildcard_rule.rs]: This is a good catch. 

I think technically it is not necessary as no other aggregates supports `*`. However, I added a check for the aggregate name and added a test for `SUM(*)` in [a05bcbe](https://github.com/apache/arrow-datafusion/pull/10066/commits/a05bcbe075cb2c015ac880b65f55e13ba22e9ac0)

**Review by andygrove — APPROVED** [review]
Thanks @alamb. This is much easier to follow.

**Pull request merged**